src/git.rs

```diff
@@ -357,3 +376,69 @@ pub fn yank(
         Ok(())
     })
 }
+
+/// Collapse the index into a single commit, archiving the current history in a snapshot branch.
+#[swirl::background_job]
+pub fn squash_index(env: &Environment) -> Result<(), PerformError> {
+    let repo = env.lock_index()?;
+    println!("Squashing the index into a single commit.");
+
+    let now = Utc::now().format("%Y-%m-%d");
+    let original_head = repo.head_oid()?;
+    let msg = format!("Collapse index into one commit\n\n\
+
+        Previous HEAD was {}, now on the `snapshot-{}` branch\n\n\
+
+        More information about this change can be found [online] and on [this issue].\n\n\
+
+        [online]: https://internals.rust-lang.org/t/cargos-crate-index-upcoming-squash-into-one-commit/8440\n\
+        [this issue]: https://github.com/rust-lang/crates-io-cargo-teams/issues/47", original_head, now);
+
+    // Create a snapshot branch of current `HEAD`.
+    repo.push(&format!("HEAD:refs/heads/snapshot-{}", now))?;
+
+    repo.squash_to_single_commit(&msg)?;
+
+    // Shell out to git because libgit2 does not currently support push leases
+
+    let key = match &repo.credentials {
+        Credentials::Ssh { key } => key,
+        Credentials::Http { .. } => {
+            return Err(String::from("squash_index: Password auth not supported").into())
+        }
+        _ => return Err(String::from("squash_index: Could not determine credentials").into()),
+    };
+
+    let temp_key_file = tempfile::Builder::new().tempfile()?;
+    let mut file = File::create(&temp_key_file)?;
+    file.write_all(key.as_bytes())?;
+    drop(file);
+
+    let checkout_path = repo.checkout_path.path();
+    let output = std::process::Command::new("git")
+        .current_dir(checkout_path)
+        .env(
+            "GIT_SSH_COMMAND",
+            format!(
+                "ssh -o StrictHostKeyChecking=accept-new -i {}",
+                temp_key_file.path().display()
+            ),
+        )
+        .args(&[
+            "push",
+            "origin",
+            "HEAD:refs/heads/master",
+            &format!("--force-with-lease=refs/heads/master:{}", original_head),
+        ])
+        .output()?;
+
+    if !output.status.success() {
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        let message = format!("Running git command failed with: {}", stderr);
+        return Err(message.into());
+    }
```

Comment:
I'm wondering... should we delete the `snapshot-` branch we just created if this fails?

Reply:
I've combined both push operations, adding `--atomic` to the command line args.

Reply:
Even better! And TIL about `--atomic`.